server/resp/src/main/java/org/infinispan/server/resp/SubscriberHandler.java

```diff
@@ -92,76 +96,78 @@ public RespRequestHandler handleRequest(ChannelHandlerContext ctx, String type,
+                        (__, t) -> {
+                           byte[] keyChannel = entry.getKey().getBytes();
+                           if (t != null) {
+                              // TODO: not sure what to do with multiple errors or if some are added and some failed..
+                              log.exceptionWhileRemovingListener(t, CharsetUtil.UTF_8.decode(ByteBuffer.wrap(keyChannel)));
+                              ctx.writeAndFlush("-ERR Failure unsubscribing client listener");
+                           } else {
+                              sendUnsubscribe(ctx, keyChannel);
+                           }
+                           iterator.remove();
+                        }));
                }
             } else {
                for (byte[] keyChannel : arguments) {
                   WrappedByteArray wrappedByteArray = new WrappedByteArray(keyChannel);
                   PubSubListener listener = specificChannelSubscribers.remove(wrappedByteArray);
                   if (listener != null) {
-                     respServer.getCache().removeListenerAsync(listener)
-                           .whenComplete((ignore, t) -> {
+                     aggregateCompletionStage.dependsOn(stageToReturn(respServer.getCache().removeListenerAsync(listener), ctx,
+                           (__, t) -> {
                               if (t != null) {
+                                 // TODO: not sure what to do with multiple errors or if some are added and some failed..
                                  log.exceptionWhileRemovingListener(t, CharsetUtil.UTF_8.decode(ByteBuffer.wrap(keyChannel)));
                                  ctx.writeAndFlush("-ERR Failure unsubscribing client listener");
                               } else {
                                  sendUnsubscribe(ctx, keyChannel);
                               }
-                           });
+                           }));
                   } else {
                      sendUnsubscribe(ctx, keyChannel);
                   }
                }
             }
-            break;
+            return aggregateCompletionStage.freeze();
          case "PING":
-            // Note we don't return the handler and just use it to handle the ping
+            // Note we don't return the handler and just use it to handle the ping - we assume stage is always complete
             handler.handleRequest(ctx, type, arguments);
             break;
          case "RESET":
-            for (Iterator<Map.Entry<WrappedByteArray, PubSubListener>> iterator = specificChannelSubscribers.entrySet().iterator(); iterator.hasNext(); ) {
-               Map.Entry<WrappedByteArray, PubSubListener> entry = iterator.next();
-               PubSubListener listener = entry.getValue();
-               respServer.getCache().removeListenerAsync(listener);
-               iterator.remove();
-               sendUnsubscribe(ctx, entry.getKey().getBytes());
-            }
```

Comment:
Why not unsubscribe anymore?

Reply:
Looking at the spec of RESET there is no reason to. https://redis.io/commands/reset/
All it says is "Aborts Pub/Sub's subscription state ([SUBSCRIBE](https://redis.io/commands/subscribe) and [PSUBSCRIBE](https://redis.io/commands/psubscribe)), when appropriate."

But I can undo this and possibly look into it in a different PR/JIRA.

Reply:
Looking at this more, I think I want to redo how the unsubscribe is done for the pub/sub anyways as I wasn't sure how I wanted to the error handling either.

Reply:
Actually looking closer, I think this is the proper thing to do as the lettuce client completely obliterates its connection.

https://github.com/lettuce-io/lettuce-core/blob/main/src/main/java/io/lettuce/core/protocol/DefaultEndpoint.java#L645

Reply:
I removed this as part of a different commit for [ISPN-14245](https://issues.redhat.com/browse/ISPN-14245)

Reply:
That is a good point about leakage, and actually there is a leak outside of this if the socket gets disconnected for example. I will add back in the removal of the listeners, however I won't call sendUnsubscribe.  I need to think of the best way to handle the client disconnect still, because that will happen with lettuce on a reset for example. I will enhance the test to verify this as well.

Reply:
Updated to remove all the listeners on QUIT/RESET and when the netty channel is closed. Also added a test to verify they are removed as expected (test couldn't use quit or reset commands as lettuce is doing some funky stuff there that I don't trust).